### PR TITLE
Picturae import

### PR DESCRIPTION
**Pull request by delarocamcas**
These are additions I made to create picturae import:
new .py files include:

specify7-ipup.py. (automatically updates local ip address in settings , and config files)
image_client/csv_purger.py (removes upload batches according to MD5 code)
image_client/picturae_config.template.py
image_client/picturae_csv_create.py. (wrangles csv files, and parses them for upload)
image_client/picturae_importer.py (inserts new records into ntobiko tables and taxon tree)
image_client/taxon_parse_utils.py (for cleaning and parsing taxonomic nomenclature)
image_client/picturae_import_utils.py (for modifying directory + operations for import lists/ dictionaries)
image_client/string_utils.py (utils for parsing and modifying strings)
image_client/sql_csv_utils.py (used for parsing generalized sql queries )
new .R files include:
- image_client/taxon_check/test_TNRS.R (checks for spelling and validity of taxon strings)
** There are still some edits in progress on some of these files **

### CODE REVIEW

File: .github/workflows/db_credentials.yml
Comment (foozleface) [inline .github/workflows/db_credentials.yml:1]: I think we were going to drop this file, as we didn't implement secrets? or no?

File: image_client/botany_importer.py [hunk 1]
```diff
@@ -59,7 +53,7 @@ def process_barcode(self, barcode, filepath_list):
             self.logger.debug(f"No barcode; skipping")
             return
         self.logger.debug(f"Barcode: {barcode}")
-        sql = f"select collectionobjectid  from collectionobject where catalognumber={barcode}"
+        sql = f"select CollectionObjectID from casbotany.collectionobject where CatalogNumber={barcode};"
```
Comment (foozleface) [inline image_client/botany_importer.py]: remove casbotany. prefix from these calls unless we need it - it makes the code more applicible to other collections if we end up re-using it.

File: image_client/botany_importer.py [hunk 2]
```diff
@@ -97,15 +95,18 @@ def build_filename_map(self, full_path):
...
+
+
+
     def create_skeleton(self, barcode):
         self.logger.info(f"Creating skeleton for barcode {barcode}")
         barcode = str(barcode).zfill(9)
         cursor = self.specify_db_connection.get_cursor()
         collecting_event_guid = uuid4()
-        sql = (f"""INSERT INTO collectingevent (
+        sql = (f"""INSERT INTO casbotany.collectingevent (
```
Comment (foozleface) [inline image_client/botany_importer.py:110]: casbotany prefix

File: image_client/botany_importer.py [hunk 3]
```diff
@@ -118,18 +119,18 @@ def create_skeleton(self, barcode):
...
-        )""")
+        );""")
         self.logger.debug(sql)
         cursor.execute(sql)
         self.specify_db_connection.commit()
 
         cursor.close()
 
-        sql = f"select CollectingEventID from collectingevent where guid='{collecting_event_guid}'"
+        sql = f"select CollectingEventID from casbotany.collectingevent where GUID='{collecting_event_guid}';"
```
Comment (foozleface) [inline image_client/botany_importer.py]: There are more, I'll stop flagging each one :)

File: image_client/botany_importer_unittest.py
Comment (foozleface) [inline image_client/botany_importer_unittest.py:1]: Classically these are in a subdirectory "tests"; let's make it that way here.

File: image_client/client_tools.py [hunk 4]
```diff
@@ -40,18 +46,40 @@ def parse_command_line():
...
 
 def main(args):
 
     if args.subcommand == 'search':
         image_client = ImageClient()
     elif args.subcommand == 'import':
-
         if args.collection == "Botany":
-             BotanyImporter()
+            # get paths here
```
Comment (foozleface) [inline image_client/client_tools.py:60]: perfect.

File: image_client/csv_purger.py
Comment (foozleface) [inline image_client/csv_purger.py:1]: Why do we purge upload batches? what's being purged? is this a standalone program? if the latter, does it make sense for it to be in its own subdirectory?

File: image_client/data_utils.py
Comment (foozleface) [inline image_client/data_utils.py:1]: Break this up per comments in previous PR

File: image_client/db_utils.py [hunk 5]
```diff
@@ -45,41 +48,39 @@ def reset_connection(self):
...
-                                                   database=self.database_name)
-            except mysql.connector.Error as err:
-                if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
+                self.cnx = mariadb.connect(user=self.database_user,
+                                           password=self.database_password,
+                                           host=self.database_host,
+                                           port=self.database_port,
+                                           database=self.database_name)
+            except mariadb.Error as err:
+                print("oh no! error")
```
Comment (foozleface) [inline image_client/db_utils.py]: print(f"MariaDB threw an error: {err}")

File: image_client/duplicates-Botany.txt
Comment (foozleface) [inline image_client/duplicates-Botany.txt:1]: do not commit

File: image_client/iz_importer_config.py
Comment (foozleface) [inline image_client/iz_importer_config.py:1]: We can have an iz_importer_config_template.py, but not the config itself. This looks like it's intended to be a template - see the pattern we use elsewhere with configs, where we have template files but not the "Real thing". this leaves us the ability to add iz_importer_config.py in the .gitignore for our test systems, and then we have have live passwords in it and not have to worry about an inadvertent commit.

File: image_client/nightly_sync.py [hunk 6]
```diff
@@ -3,12 +3,17 @@
 from db_utils import DbUtils
 import botany_importer_config
 import ich_importer_config
+<<<<<<< HEAD
```
Comment (foozleface) [inline image_client/nightly_sync.py]: fix per previous PR comments

File: image_client/nightly_sync.py [hunk 7]
```diff
@@ -65,12 +70,17 @@ def do_import(collection_name,specify_db_connection):
                 print(f"   internal filename: {internal_filename} redacted: {redacted}", file=sys.stderr, flush=True)
                 next_record = True
             except Exception as e:
-                print(f"Error, probably sql {e}", file=sys.stderr, flush=True)
+                print(f"Error, probably sql: \"{e}\"", file=sys.stderr, flush=True)
                 print(f"exception type: {type(e).__name__}", file=sys.stderr, flush=True)
 
+<<<<<<< HEAD
```
Comment (foozleface) [inline image_client/nightly_sync.py]: fix

File: image_client/nightly_sync.py [hunk 8]
```diff
@@ -93,13 +103,17 @@ def main():
             ich_importer_config.SPECIFY_DATABASE_PORT,
             ich_importer_config.SPECIFY_DATABASE_HOST,
             ich_importer_config.SPECIFY_DATABASE)
-    do_import(collection_name,specify_db_connection)
+    do_sync(collection_name, specify_db_connection)
 
 
 if __name__ == '__main__':
+<<<<<<< HEAD
```
Comment (foozleface) [inline image_client/nightly_sync.py]: fix

File: image_client/picturae_img/PIC_2023-06-28/.hidden_.hidden_.hidden_.DS_Store
Comment (foozleface) [inline image_client/picturae_img/PIC_2023-06-28/.hidden_.hidden_.hidden_.DS_Store:1]: do not commit.

File: image_client/picturae_importer.py
Comment (foozleface) [inline image_client/picturae_importer.py:1]: Haven't read in detail yet - when you fix the other PR comments here, I'll read it. however, 1000 line classes are usually awkward to use - see if there's a natural way to break this up that makes sense to you. if you don't see one, we'll go over it when we do the next round

File: image_client/specify_db.py [hunk 9]
```diff
@@ -2,10 +2,11 @@
...
 class SpecifyDb(DbUtils):
-    def __init__(self,db_config_class):
+    def __init__(self, db_config_class):
         self.specify_db_connection = super().__init__(
             db_config_class.USER,
             db_config_class.PASSWORD,
             db_config_class.SPECIFY_DATABASE_PORT,
             db_config_class.SPECIFY_DATABASE_HOST,
             db_config_class.SPECIFY_DATABASE)
+        print("initializing that thing")
```
Comment (foozleface) [inline image_client/specify_db.py]: let's be a little more specific here, and it should also probably be a logger.debug or a logger.info.

File: image_client/taxon_check/.readme.txt
Comment (foozleface) [inline image_client/taxon_check/.readme.txt:1]: probably shoujdn't start with a "."

File: image_client/taxon_check/output.txt
Comment (foozleface) [inline image_client/taxon_check/output.txt:1]: do not commit

File: image_client/test_csv_purge_sql/.readme.txt
Comment (foozleface) [inline image_client/test_csv_purge_sql/.readme.txt:1]: dont' start with .

File: image_client/unit_test_picturae.py
Comment (foozleface) [inline image_client/unit_test_picturae.py:1]: put in test dir

File: server.py [hunk 10]
```diff
@@ -189,7 +189,7 @@ def resolve_file(filename, collection, type, scale):
     """Inspect the request object to determine the file being requested.
     If the request is for a thumbnail and it has not been generated, do
     so before returning.
-
+accession_copy
```
Comment (foozleface) [inline server.py]: is this a typo?

**Review by foozleface — CHANGES_REQUESTED** [review]
I haven't fully read the big files you wrote yet; I'll do that when you've addressed these issues. Generally, it's a good idea to look at the contents of the PR before having someone else review them; there's a lot here that you would have caught with a quick scan.

File: image_client/db_utils.py (hunk 5 above)
Comment (delarocamcas) [reply image_client/db_utils.py]: my bad, leftover silly print statement from debugging,  I'll replace with the above print statement haha.

File: image_client/picturae_importer.py
Comment (delarocamcas) [reply image_client/picturae_importer.py:1]: I'll brain storm a bit before the next PR, see what I can do to make it shorter.

File: image_client/Picturae_DDL.sql
Comment (foozleface) [inline image_client/Picturae_DDL.sql:1]: Which database is being used here? images, specify, or a new one that you're using for this purpose?

File: image_client/client_tools.py [hunk 11]
```diff
@@ -68,13 +68,15 @@ def main(args):
             BotanyImporter(paths=paths, config=botany_importer_config)
         elif args.collection == 'Botany_PIC':
             #get paths here
+            date_override = args.date
+            if date_override is not None:
+                rewrite_config_date(date_override, filepath="image_client/picturae_config.py")
```
Comment (foozleface) [inline image_client/client_tools.py]: picturae_config.py is a python file; you don't need to parse it; Just import it and then use the variables therein.

File: image_client/PIC_upload_purger/csv_purger.py
Comment (foozleface) [inline image_client/PIC_upload_purger/csv_purger.py:1]: Okay; I htink I get this. Essentially this is an "undo" for a bad import. Let's rename it? undo_csv_import.py or something like that? Is that correct?

**Pull request closed without merging**